src/experiments.js

```diff
@@ -54,6 +58,109 @@ export function isCanary(win) {
+ *   2. Crypto isn't available,
+ *   3. The experiment is not specified in the experiment token
+ *   4. The experiment has expired.
+ * The promise returned by this function will reject with an error if:
+ *   1. The experiment meta tag is present without a token
+ *   2. The token is malformed (e.g. non-existant version number)
+ *   3. The token is not for this origin
+ *   4. The experiments data was not signed with our private key
+ * @param {!Window} win
+ * @param {string} experimentId
+ * @param {!Object} opt_publicJwk Used for testing only.
+ * @return {!Promise<boolean>}
+ */
+export function isExperimentOnForOriginTrial(win, experimentId, opt_publicJwk) {
+  const crypto = cryptoFor(win);
+  const meta =
+      win.document.head.querySelector('meta[name="amp-experiment-token"]');
+  if (!meta || !crypto.isCryptoAvailable()) {
+    return Promise.resolve(false);
+  }
+  const token = meta.getAttribute('content');
+  if (!token) {
+    return Promise.reject(new Error('Unable to read experiments token'));
+  }
+  /**
+   * token = encode64(version + length + config + sign(config, private_key))
+   * version = 1 byte version of the token format (starting at 0x0)
+   */
+  let current = 0;
+  const bytes = stringToBytes(atob(token));
+  const version = bytes[current];
+  if (version !== 0) {
+    // Unrecognized version number
+    const error =
+        new Error(`Unrecognized experiments token version: ${version}`);
+    return Promise.reject(error);
+  }
+  current++;
+ /**
+  * Version 0:
+  * length = 4 bytes representing number of bytes in config
+  * config = string containing the experiment ID, origin URL, etc.
+  */
+  const bytesForConfigSize = 4;
+  const configLen =
+      bytesToUInt32(bytes.subarray(current, current + bytesForConfigSize));
+  current += bytesForConfigSize;
+  if (configLen > bytes.length - current) {
+    return Promise.reject(
+        new Error('Specified len extends past end of buffer'));
+  }
+  const configBytes = bytes.subarray(current, current + configLen);
+  current += configLen;
+  const signatureBytes = bytes.subarray(current);
+
+  // TODO(kmh287, choumx) fill in real public key
+  const publicJwk = opt_publicJwk || {};
+
+  return crypto.importPublicKey('experiments', publicJwk).then(keyInfo => {
+    return crypto.verifySignature(configBytes, signatureBytes, keyInfo);
+  }).then(verified => {
+    if (!verified) {
+      throw new Error('Failed to verify config signature');
+    }
+    const configStr = utf8DecodeSync(configBytes);
+    const config = JSON.parse(configStr);
+
+    const approvedOrigin = config['origin'];
+    const url = ampdocServiceFor(win).getAmpDoc().getUrl();
+    const sourceOrigin = getSourceOrigin(url);
+    if (approvedOrigin !== sourceOrigin) {
+      throw new Error('Config does not match current origin');
+    }
+
+    const experiments = config['experiments'];
+    const experiment = experiments[experimentId];
+    if (!experiment) {
+      return false;
+    }
```

Comment:
Then we should probably run this at page load. Otherwise we will need to apply this process to every token on the page. We could do it once lazily, but then we could still have a promise that takes a very long time to resolve.

WDYT?